src/test/Serialize_test.cpp

```diff
@@ -0,0 +1,224 @@
+                Issue(to_currency("USD"), *amountAccount),
+                    123400000));
+        }
+        auto const sendMaxAccount = parseBase58<AccountID>(
+            "razqQKzJRdB4UxFPWf5NEpEG3WMkmwgcXA");
+        if (BEAST_EXPECT(sendMaxAccount))
+        {
+            // Need a variable here instead of a literal to make
+            // some compilers happy
+            const std::uint64_t mantissa = 5678900000000000LLU;
+            BEAST_EXPECT(obj[sfSendMax] == STAmount(
+                Issue(to_currency("CNY"), *sendMaxAccount),
+                mantissa, -4));
+        }
+    }
+
+    void testParseJson()
+    {
+        testcase("ParseJson");
+
+        auto& testTx = getKnownTx();
+        auto json = parseJson(testTx.JsonText);
+        BEAST_EXPECT(json);
+
+        // Don't need to test the `JsonReader`, just a case to ensure the wrapper
+        // handles failures
+        json = parseJson("{ asjlfkjs");
+        BEAST_EXPECT(!json);
+    }
+
+    void testMakeObject()
+    {
+        testcase("Make Object");
+
+        auto& testTx = getKnownTx();
+        auto json = parseJson(testTx.JsonText);
+        auto obj = makeObject(json);
+        if (BEAST_EXPECT(obj))
+        {
+            ripple::STTx tx{ std::move(*obj) };
+            verifyKnownTx(tx);
+        }
+    }
+
+    void testSerialize()
+    {
+        testcase("Serialize");
+
+        auto test = [&](TestItem const& testItem)
+        {
+            auto json = parseJson(testItem.JsonText);
+            auto obj = makeObject(json);
+            if (BEAST_EXPECT(obj))
+            {
+                auto serialized = serialize::serialize(*obj);
+                BEAST_EXPECT(testItem.SerializedText == serialized);
+            }
+        };
+
+        test(getKnownTx());
+        test(getKnownMetadata());
+        //test(getKnownLedger());
+    }
+
+    void testDeserialize()
+    {
+        testcase("Deserialize");
+
+        auto test = [&](TestItem const& testItem,
+            std::function<bool(ripple::STObject&,
+                Json::Value const& known)> extra = nullptr)
+        {
+            try
+            {
+                auto obj = deserialize(testItem.SerializedText);
+
+                if (BEAST_EXPECT(obj))
+                {
+                    auto check = true;
+                    auto known = parseJson(testItem.JsonText);
+                    if (extra)
+                        check = extra(*obj, known);
+                    if (check)
+                    {
+                        BEAST_EXPECT(obj->getJson(0) == known);
+                    }
+                }
+            }
+            catch (...)
+            {
+                fail();
+            }
+        };
+
+        test(getKnownTx(), [&](auto& obj, auto const& known)
+        {
+            ripple::STTx tx{ std::move(obj) };
+            this->verifyKnownTx(tx);
+            this->BEAST_EXPECT(tx.getJson(0) == known);
+            return false;
+        });
```

Comment:
The lambda here is the optional `extra` param to `test`. `test` itself is called three (now) times.